### DISCUSSION TITLE
How to pass list in query parameters?

### DISCUSSION THREAD

**Discussion by Yolley** [fastapi/fastapi#8225] [Questions] [answered]
**Description**

I need to support several query parameters with same name in GET route. Typical request looks like this:
http://localhost/item?num=1&num=2

I configured a route
````
@app.get("/item", content_type=UJSONResponse)
async def get_part(num: list):
    found = False
    for i in num:
         if i in some_data:
             found = True
             break
    return {"found": found}
````
The idea is, that I can pass several numbers to check, if they exist in some_data. In my opinion, it is not a good idea to get numbers from body here, because it is a simple GET request, but in current version application expects argument "num" in body, so on my request (written above) I receive a response

`{"detail":[{"loc":["body","num"],"msg":"field required","type":"value_error.missing"}]}`

As I know, it is not restricted to pass several query parameters with the same name in an HTTP request by any specifications, so I would like to ask, is it possible to configure a route in FastAPI, which will be able to parse several query parameters with the same name?

**Comment by tiangolo** [comment, marked as answer]
It is supported (and tested), but not documented yet, but you can just use standard Python types (a `List`) :)

```Python
from typing import List

from fastapi import FastAPI, Query

app = FastAPI()
@app.get("/items/")
def read_items(q: List[int] = Query(None)):
    return {"q": q}
```

**Comment by tiangolo** [comment]
It is now properly documented :memo: :tada: 

For query parameters: https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#query-parameter-list-multiple-values

And for duplicate headers: https://fastapi.tiangolo.com/tutorial/header-params/#duplicate-headers

**Reply by epicwhale** [reply]
The example that is documented in the docs with Annotations, is throwing an AssertionError when I try it. I've reported the issue here: https://github.com/tiangolo/fastapi/discussions/9283

**Comment by tiangolo** [comment]
I guess this is solved now, so I'll close the issue. But feel free to add more comments or create new issues.

**Reply by rsbabcock** [reply]
@tiangolo Do you have any best practices on what should be a reasonable limit on how many items in a list to pass into a list type parameter? I.e have a need to filter on 100+ records and I was able to pass in 1500 as a list but couldn't find documentation specific to fast api if that is recommended or not.

**Comment by neilpanchal** [comment]
Is it possible to do:
`http://localhost/item?num=1,2,3,4,5,6`

instead of:
`http://localhost/item?num=1&num=2&num=3&num=4&num=5&num=6`?

**Comment by dmontagu** [comment]
@neilpanchal You could do this by adding a custom dependency that expected a string-value query parameter, then splits the string on `,` and converts the result to ints. But obviously that's a little less ergonomic.

I think the main reason why this isn't easier is that FastAPI is basically just deferring to starlette for query parameter string extraction, then pydantic for parsing.

So starlette gives query parameter `num="1,2,3,4,5,6"`, and pydantic doesn't know how to parse `"1,2,3,4,5,6"` into a list of ints.

Starlette *does* have built-in support for repeated values of the same query parameter, which is why the more verbose approach does work out of the box.

**Comment by falkben** [comment]
here's an example dependency that I wrote to handle this sort of thing

I wanted to be able to handle normal query arg lists as well as these awful "string" lists that are wrapped in square brackets and sometimes contain double quotes and spaces that were somehow a part of our existing API.

```py
def parse_list(names: List[str] = Query(None)) -> Optional[List]:
    """
    accepts strings formatted as lists with square brackets
    names can be in the format
    "[bob,jeff,greg]" or '["bob","jeff","greg"]'
    """
    def remove_prefix(text: str, prefix: str):
        return text[text.startswith(prefix) and len(prefix):]

    def remove_postfix(text: str, postfix: str):
        if text.endswith(postfix):
            text = text[:-len(postfix)]
        return text

    if names is None:
        return

    # we already have a list, we can return
    if len(names) > 1:
        return names

    # if we don't start with a "[" and end with "]" it's just a normal entry
    flat_names = names[0]
    if not flat_names.startswith("[") and not flat_names.endswith("]"):
        return names

    flat_names = remove_prefix(flat_names, "[")
    flat_names = remove_postfix(flat_names, "]")

    names_list = flat_names.split(",")
    names_list = [remove_prefix(n.strip(), "\"") for n in names_list]
    names_list = [remove_postfix(n.strip(), "\"") for n in names_list]

    return names_list


@app.get("/hello_list")
def hello_list(names: List[str] = Depends(parse_list)):
    """ list param method """

    if names is not None:
        return StreamingResponse((f"Hello {name}" for name in names))
    else:
        return {"message": "no names"}
```

**Comment by stratosgear** [comment]
Hey @falkben (or any one else in the thread),

This is a great solution and works great.

But I had to refactor my method signature from:

```
@app.get("/hello_list")
def hello_list(names: List[str] = Query(None, description="List of names to greet")):
    """ list param method """
```

to

```
@app.get("/hello_list")
def hello_list(names: List[str] = Depends(parse_list)):
    """ list param method """
```

and I lost the nice documentation provided by the 'description' field.

Any idea how I could combine `Query` with `Depends`?  Should I open another issue for greater visibility?

**Comment by falkben** [comment]
You should be able to move the description into the Depends method.  So in the example, the depends method signature would look like this:

```python
def parse_list(names: List[str] = Query(None, description="List of names to greet")) -> Optional[List]:
```

**Comment by stratosgear** [comment]
Hmmm, that is interesting, but that locks you into using the same description to all query params that you might want to check.

But wait, now that I am looking at it, this looks like it is tied to one query param name only, and the same 'parse_list` method cannot be reused for another param....? :(

Seems like, I'll have to take a closer look... 

**[EDITED]**: to say that your suggestion is indeed working, for one param that appears as a list, but I cannot add a second one.

**Comment by sylann** [comment]
Is there a "proper" way to handle lists using the following format: `/foo?items[]=1&items[]=2`?

This format is used by the axios library which is well known in javascript.

Right now, I use the following trick:
```py
@app.get("/foo")
async def foo(items: List[int] = Query(None, alias="items[]")):
    ...
```
The problem is that it only handles this syntax in this case and passing this won't work: `/foo?items=1&items=2`.

One could argue about the fact that the API decides what format to use I guess. But still, it would be cool to handle both syntax as they are pretty close.

**Reply by mayurpande** [reply]
Did you ever find a solution to this? As vuetify2 multi-sort uses the `/foo?items[]=1&items[]=2`, I am going to use the trick you have said, but it would be nice to get an "official" method.

**Comment by allenhumphreys** [comment]
In case it helps anyone, I used a middleware to turn comma-delimited query parameter strings into repeated query parameters which work well with the out of the box/Pydantic list parsing. Pure-python in 3.7 or later (maybe lower?)


```python3
import fastapi

from starlette.types import ASGIApp, Scope, Receive, Send
from urllib.parse import parse_qs as parse_query_string
from urllib.parse import urlencode as encode_query_string

class QueryStringFlatteningMiddleware:
    def __init__(self, app: ASGIApp) -> None:
        self.app = app

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        query_string = scope.get("query_string", None).decode()
        if scope["type"] == "http" and query_string:
            parsed = parse_query_string(query_string)
            flattened = {}
            for name, values in parsed.items():
                all_values = []
                for value in values:
                    all_values.extend(value.split(","))

                flattened[name] = all_values

            # doseq: Turn lists into repeated parameters, which is better for FastAPI
            scope["query_string"] = encode_query_string(flattened, doseq=True).encode("utf-8")

            await self.app(scope, receive, send)
        else:
            await self.app(scope, receive, send)

app = fastapi.FastAPI()
app.add_middleware(QueryStringFlatteningMiddleware)
```

Converts `?state=COMPLETED,ENQUEUED` -> `?state=COMPLETED&state=ENQUEUED`

**note:** This code hasn't gone through any kind of hardening, it's hot off the presses so to speak.

**Comment by igorgbianchi** [comment]
> In case it helps anyone, I used a middleware to turn comma-delimited query parameter strings into repeated query parameters which work well with the out of the box/Pydantic list parsing. Pure-python in 3.7 or later (maybe lower?)
> 
> ```python
> import fastapi
> 
> from starlette.types import ASGIApp, Scope, Receive, Send
> from urllib.parse import parse_qs as parse_query_string
> from urllib.parse import urlencode as encode_query_string
> 
> class QueryStringFlatteningMiddleware:
>     def __init__(self, app: ASGIApp) -> None:
>         self.app = app
> 
>     async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
>         query_string = scope.get("query_string", None).decode()
>         if scope["type"] == "http" and query_string:
>             parsed = parse_query_string(query_string)
>             flattened = {}
>             for name, values in parsed.items():
>                 all_values = []
>                 for value in values:
>                     all_values.extend(value.split(","))
> 
>                 flattened[name] = all_values
> 
>             # doseq: Turn lists into repeated parameters, which is better for FastAPI
>             scope["query_string"] = encode_query_string(flattened, doseq=True).encode("utf-8")
> 
>             await self.app(scope, receive, send)
>         else:
>             await self.app(scope, receive, send)
> 
> app = fastapi.FastAPI()
> app.add_middleware(QueryStringFlatteningMiddleware)
> ```
> 
> Converts `?state=COMPLETED,ENQUEUED` -> `?state=COMPLETED&state=ENQUEUED`
> 
> **note:** This code hasn't gone through any kind of hardening, it's hot off the presses so to speak.


Thanks!
Would be useful if FastAPI had this as default.

**Comment by GabrieleMazzola** [comment]
> here's an example dependency that I wrote to handle this sort of thing
> 
> I wanted to be able to handle normal query arg lists as well as these awful "string" lists that are wrapped in square brackets and sometimes contain double quotes and spaces that were somehow a part of our existing API.
> 
> ```python
> def parse_list(names: List[str] = Query(None)) -> Optional[List]:
>     """
>     accepts strings formatted as lists with square brackets
>     names can be in the format
>     "[bob,jeff,greg]" or '["bob","jeff","greg"]'
>     """
>     def remove_prefix(text: str, prefix: str):
>         return text[text.startswith(prefix) and len(prefix):]
> 
>     def remove_postfix(text: str, postfix: str):
>         if text.endswith(postfix):
>             text = text[:-len(postfix)]
>         return text
> 
>     if names is None:
>         return
> 
>     # we already have a list, we can return
>     if len(names) > 1:
>         return names
> 
>     # if we don't start with a "[" and end with "]" it's just a normal entry
>     flat_names = names[0]
>     if not flat_names.startswith("[") and not flat_names.endswith("]"):
>         return names
> 
>     flat_names = remove_prefix(flat_names, "[")
>     flat_names = remove_postfix(flat_names, "]")
> 
>     names_list = flat_names.split(",")
>     names_list = [remove_prefix(n.strip(), "\"") for n in names_list]
>     names_list = [remove_postfix(n.strip(), "\"") for n in names_list]
> 
>     return names_list
> 
> 
> @app.get("/hello_list")
> def hello_list(names: List[str] = Depends(parse_list)):
>     """ list param method """
> 
>     if names is not None:
>         return StreamingResponse((f"Hello {name}" for name in names))
>     else:
>         return {"message": "no names"}
> ```

I've done a little modification to this in order to make the code reusable across different usecases. This is the code:
```python

from typing import List, Optional, Type
from fastapi import Query

def parse_list(query: Query, class_type: Type):
    def inner_parse(elements: List[str] = query) -> Optional[List[class_type]]:
        def remove_prefix(text: str, prefix: str):
            return text[text.startswith(prefix) and len(prefix):]

        def remove_postfix(text: str, postfix: str):
            if text.endswith(postfix):
                text = text[:-len(postfix)]
            return text

        if query.default != Ellipsis and not elements:
            return query.default

        if len(elements) > 1:
            elements_list = elements
        else:
            flat_elements = elements[0]
            if flat_elements.startswith("["):
                flat_elements = remove_prefix(flat_elements, "[")

            if flat_elements.endswith("]"):
                flat_elements = remove_postfix(flat_elements, "]")

            if query.default != Ellipsis and not flat_elements:
                return query.default

            elements_list = flat_elements.split(",")
            elements_list = [remove_prefix(n.strip(), "\"") for n in elements_list]
            elements_list = [remove_postfix(n.strip(), "\"") for n in elements_list]

        errors = {}
        results = []
        for idx, el in enumerate(elements_list):
            try:
                results.append(class_type(el))
            except ValueError as e:
                errors[idx] = repr(e)
        if errors:
            raise Exception(f"Could not parse elements: {errors}")
        else:
            return results
    return inner_parse
```
The idea here is that I can call the ```parse_list``` function passing the Query element that I need, thus making the code more reusable. 

For example, I can create an Enum like this:
```python
class SimpleEnum(str, Enum):
    aaa = 'aaa'
    bbb = 'bbb'
    ccc = 'ccc'
```
and then use the ```parse_list``` function like this:
```python
parsed_enums: List[SimpleEnum] = Depends(
    parse_list(query=Query([], alias="enums", description="Some description"), class_type=SimpleEnum)
)
```
This way, the Query object can be defined in the proper router.

Disclaimer: I did not spend time making this code perfectly robust, but it works just fine for what I needed. I hope this can help.

**Reply by BloodyRain2k** [reply]
Thanks for sharing, it's been really useful to me.

I've made some slight changes to reduce (in my case at least) the amount of info needed to be provided.

For an endpoint where I return a query result and needed a filter for which fields to return, while ensuring `id` is always included, I can now define that with just:
`filter:Optional[List[str]] = fa.Depends(parse_query_list('filter', str, ['id']))`
"filter" connects the alias so the query parses the right data, `str` should be obvious and `['id']` makes the `id` field always included.

```
def parse_query_list(query:Union[Query, str], class_type:Type, ensure:Optional[List[str]] = None):
    """
    `query`: If provided a `str`, will turn it into a simple `Query([], alias=query)`.
    
    `ensure`: If provided will be included in result set regardless of input.
    
    Example::
    
        @app.get('/',
            filter:List[str] = Depends(
                parse_query_list(
                    class_type = str,
                    query = Query(
                            [], alias="filter", description="Some description"
                    ),
                )
            )
        )
    """
    if isinstance(query, str):
        query = Query([], alias=query)
    
    def inner_parse(elements: List[str] = query) -> Optional[List[class_type]]:
        def remove_prefix(text: str, prefix: str):
            return text[text.startswith(prefix) and len(prefix):]

        def remove_postfix(text: str, postfix: str):
            if text.endswith(postfix):
                text = text[:-len(postfix)]
            return text

        if query.default != Ellipsis and not elements:
            return query.default

        if len(elements) > 1:
            elements_list = elements
        else:
            flat_elements = elements[0]
            if flat_elements.startswith("["):
                flat_elements = remove_prefix(flat_elements, "[")

            if flat_elements.endswith("]"):
                flat_elements = remove_postfix(flat_elements, "]")

            if query.default != Ellipsis and not flat_elements:
                return query.default

            elements_list = flat_elements.split(",")
            elements_list = [remove_prefix(n.strip(), "\"") for n in elements_list]
            elements_list = [remove_postfix(n.strip(), "\"") for n in elements_list]

        errors = {}
        results = [*ensure] if ensure else []
        for idx, el in enumerate(elements_list):
            try:
                results.append(class_type(el))
            except ValueError as e:
                errors[idx] = repr(e)
        if errors:
            raise Exception(f"Could not parse elements: {errors}")
        else:
            return results
    return inner_parse
```

**Comment by avylove** [comment]
A simpler implementation of the middleware solution from @allenhumphreys. Works for me, use at your own risk.
Like the original, Converts `?state=COMPLETED,ENQUEUED` to `?state=COMPLETED&state=ENQUEUED`.

``` python
from urllib.parse import urlencode
from fastapi import Request

@app.middleware("http")
async def flatten_query_string_lists(request: Request, call_next):

    flattened = []
    for key, value in request.query_params.multi_items():
        flattened.extend((key, entry) for entry in value.split(','))

    request.scope["query_string"] = urlencode(flattened, doseq=True).encode("utf-8")

    return await call_next(request)
```

**Comment by warnus** [comment]
@avylove Thanks for the good work. But It could be a problem cause splitting all strings. If I get Query parameter like "Hello, World". This is just long string not List parameter, but the code will split the sentence.

**Comment by dotX12** [comment]
Possible solution I'm using.

Handler: https://github.com/dotX12/pyfa-converter/blob/master/examples/main.py#L50
Model: https://github.com/dotX12/pyfa-converter/blob/master/examples/models.py#L51

Result:
![image](https://user-images.githubusercontent.com/64792903/172029951-b1c09e43-3b61-4a8e-b67e-d529f8ebd449.png)

**Comment by xoelop** [comment]
@GabrieleMazzola I liked your function! But to make it work (on FastAPI 0.78) I had to set the `alias` parameter of the `Query` that I pass to `parse_list` have the same name as the query parameter, `parsed_enums` in the example you showed

```
parsed_enums: List[SimpleEnum] = Depends(
    parse_list(query=Query([], alias="parsed_enums", description="Some description"), class_type=SimpleEnum)
)
```

Super useful anyway to not have to declare a new dependency per parameter, thanks!

**Comment by gtors** [comment]
Here is my solution (py3.10), I hope it will be useful for someone:
```py
class TypeParametersMemoizer(type):
    _generics_cache = weakref.WeakValueDictionary()

    def __getitem__(cls, typeparams):

        # prevent duplication of generic types
        if typeparams in cls._generics_cache:
            return cls._generics_cache[typeparams]

        # middleware class for holding type parameters
        class TypeParamsWrapper(cls):
            __type_parameters__ = typeparams if isinstance(typeparams, tuple) else (typeparams,)

            @classmethod
            def _get_type_parameters(cls):
                return cls.__type_parameters__

        return types.GenericAlias(TypeParamsWrapper, typeparams)


class CommaSeparatedList(list, metaclass=TypeParametersMemoizer):
    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v: str | list[str]):
        if isinstance(v, str):
            v = v.split(",")
        else:
            v = list(itertools.chain.from_iterable((x.split(",") for x in v)))
        params = cls._get_type_parameters()
        return pydantic.parse_obj_as(list[params], list(map(str.strip, v)))

    @classmethod
    def _get_type_parameters(cls):
        raise NotImplementedError("should be overridden in metaclass")
```
Usage:
```
@app.get("/items")
def get_items(ids: CommaSeparatedList[int] = Query(...)):
     ...
```

pros:
- it's works (at least with py 3.10)
- openapi works as expected
- supports both formats `?param=FOO&param=BAR` & `?param=FOO,BAR`

cons:
- complex
- looks like black magic (may stop working after a new version of python)

**Reply by feluelle** [reply]
Thanks, @gtors! Do you have an updated version supporting pydantic v2? That would be awesome 😬

**Reply by gtors** [reply]
An updated version is coming soon :)

**Reply by feluelle** [reply]
Sharing my str-only `CommaSeparatedList` in case it is useful to anyone. :)

```python
@dataclass
class MyAfterValidator:
    func: Callable[[Any], Any]

    def __get_pydantic_core_schema__(self, source_type: Any, handler: GetCoreSchemaHandler) -> CoreSchema:
        return core_schema.no_info_after_validator_function(self.func, handler(source_type))

CommaSeparatedList = Annotated[str, MyAfterValidator(lambda v: v.split(","))]
```

**Reply by gtors** [reply]
```python
import itertools
import types
import weakref
from typing import Any

from fastapi import FastAPI, Query
from pydantic import BaseModel, GetCoreSchemaHandler, TypeAdapter
from pydantic_core import CoreSchema, core_schema


class TypeParametersMemoizer(type):
    _generics_cache = weakref.WeakValueDictionary()

    def __getitem__(cls, typeparams):
        # prevent duplication of generic types
        if typeparams in cls._generics_cache:
            return cls._generics_cache[typeparams]

        # middleware class for holding type parameters
        class TypeParamsWrapper(cls):
            __type_parameters__ = (
                typeparams if isinstance(typeparams, tuple) else (typeparams,)
            )
            __type_adapter__ = TypeAdapter(list[__type_parameters__])

            @classmethod
            def _get_type_parameters(cls):
                return cls.__type_parameters__

            @classmethod
            def _get_type_adapter(cls):
                return cls.__type_adapter__

        wrapper = cls._generics_cache[typeparams] = types.GenericAlias(TypeParamsWrapper, typeparams)
        return wrapper


class CommaSeparatedList(list, metaclass=TypeParametersMemoizer):
    @classmethod
    def __get_pydantic_core_schema__(
        cls, _source_type: Any, handler: GetCoreSchemaHandler
    ) -> CoreSchema:
        return core_schema.no_info_before_validator_function(
            cls.validate, handler(list[cls._get_type_parameters()])
        )

    @classmethod
    def validate(cls, v: Any):
        adapter = cls._get_type_adapter()
        if isinstance(v, str):
            v = map(str.strip, v.split(","))
        elif isinstance(v, list) and all(isinstance(x, str) for x in v):
            v = map(str.strip, itertools.chain.from_iterable(x.split(",") for x in v))
        return adapter.validate_python(v)

    @classmethod
    def _get_type_parameters(cls):
        raise NotImplementedError("should be overridden in metaclass")

    @classmethod
    def _get_type_adapter(cls) -> TypeAdapter:
        raise NotImplementedError("should be overridden in metaclass")


# ------------------------------------------------------------------------------


app = FastAPI()


@app.post("/foo")
async def foo(foo: CommaSeparatedList[int] = Query(...)) -> CommaSeparatedList[int]:
    print(foo)
    return foo


if __name__ == "__main__":
    foo = TypeAdapter(CommaSeparatedList[int])
    assert foo.validate_python("1,2,3") == [1, 2, 3]
    assert foo.validate_python(["1,2,3", "4,5,6"]) == [1, 2, 3, 4, 5, 6]
    assert foo.validate_python([1, 2, 3]) == [1, 2, 3]
```

As promised, here is an updated version

**Reply by jeonghanjoo** [reply]
@gtors Great! Appreciate so much! 
There is a little minor thing to fix

In `__get_pydantic_core_schema__`  of `CommaSeparatedList` , `cls.validate, handler(list[cls._get_type_parameters()]` needs `)` closing.

**Comment by manorlh** [comment]
what about using a query param with a separated string by a comma? 
for example:
/foo?id=x,y,z
expected behavior to get it in fast API like [x,y,z] currently we get it as a List with one string which is "x,y,z"

is there planning to support it? or is there any patch code for it?

**Comment by dotX12** [comment]
> what about using a query param with a separated string by a comma? for example: /foo?id=x,y,z expected behavior to get it in fast API like [x,y,z] currently we get it as a List with one string which is "x,y,z"
> 
> is there planning to support it? or is there any patch code for it?

foo: List[str] = Query(…) is 
/users?foo=a&foo=b&foo=c

**Comment by biradar73383** [comment]
**how to pass list in path parameter**

from typing import Union

from fastapi import FastAPI, Path, Query
from typing_extensions import Annotated

app = FastAPI()


@app.get("/items/{item_ids}")
def read_items(item_ids: list):
    return {"item_ids": item_ids}

 I'm trying  to pass list in path parameter its throwing error,  is it possible to pass list in path parameter

help me with this @tiangolo @stratosgear @Yolley @falkben @epicwhale @gtors @GabrieleMazzola
 @sylann

**Comment by arthuriantech** [comment]
Python 3.12:

```python
type Csv[T] = Annotated[list[T], BeforeValidator(lambda x: x.split(','), Query()]

@router.get('/test')
async def test(x: Csv[int]):
    print('x:', x)
```

**Reply by flo-at** [reply]
That works great but the OpenAPI schema doesn't tell it's an array anymore unfortunately.
Also, the [query parameter style](https://swagger.io/docs/specification/serialization/) should have `explode` set as `false` for comma-separated lists.
Any idea how to fix this?

**Comment by martincpt** [comment]
Has anyone been able to parse multi-values inside a BaseModel?

```python
class MyParams:
    status: Annotated[list[str] | None, Query()] = None
    start_date: datetime | None = None
    end_date: datetime | None = None

def my_endpoint(params: MyParams = Depends()) -> dict:
    ...
```

I was able to receive multi-values when they were passed directly as endpoint arguments, but not when using Query inside a parameter schema.

**Comment by bersena911** [comment]
There is not a good solution here. those answers are not generalized.

there should be a method that will do something like this:

```
first_list: list[str] = Depends(parse_list),
second_list: list[str] = Depends(parse_list),
```
and endpoint will accept both first_list and second_list.

in the answers above it is hardcoded to be only working for "names" param, some tricks can be done to pass name of param in this parse_list function but thats hacky way. it should be extracted from the name of a parameter itself.

**Comment by SDAravind** [comment]
Use field serializer for required fields

```python
@field_serializer(
        "field1", "field2", "field3", mode="plain", when_used="always"
    )
def str_to_list(self, values):
    return values.split(",") if values else values
  ```

**Comment by rrakso** [comment]
### Two-word query param name
Is it possible to have a "two word" query name, something like:  `http://localhost/item?num=1,2,3,4,5,6&asset-id`? Does anyone know how to achieve this with FastAPI?

**Reply by rrakso** [reply]
There is something: https://fastapi.tiangolo.com/tutorial/header-params/#automatic-conversion but this is realted to the `Headers` and not the `Query`.

**Reply by rrakso** [reply]
Hehe! Got it! https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#alias-parameters 😅

```Python
from typing import Annotated

from fastapi import FastAPI, Query

app = FastAPI()


@app.get("/items/")
async def read_items(q: Annotated[str | None, Query(alias="item-query")] = None):
    results = {"items": [{"item_id": "Foo"}, {"item_id": "Bar"}]}
    if q:
        results.update({"q": q})
    return results
```
